### ISSUE TITLE
Grad cam output for batch of images

### ISSUE THREAD

**Issue by diksham-flwls** [jacobgil/pytorch-grad-cam#520]
I would like to use grad cam for my model which accepts as input a tuple. The first elem in the tuple is a tensor and the second elem is a list of tensors corresponding to a list of frames from a video.

```
class org_model:
...
def forward(self, input):     # list of tensors
  ...
```

As as workaround I created a new model that takes as input a stack of tensors and then converts it into a list of tensors internally before calling the original model. 

```
class new_model:
  def __init__(self, model):
    self.model = model
...
  def forward(self, input):
    modified_input = input.reshape(N, T, 3, 224, 224)
    list_input = [ [frame for frame in window ] for window in modified_input]
    self.model(list_input)
```

```
m = org_model()
toy_model = new_model(m)
cam = GradCAM(model=toy_model, target_layers=target_layers)
grayscale_cam = cam(input_tensor=input_tensor, targets=targets)

```

However the output returned by cam is a single image instead of a batch of masks. I was expecting the output to be the same size as the input tensor. Could you please help me understand whats going on here?

**Comment by diksham-flwls** [comment]
A follow up question, what should be the `targets` in this case? Since the output is just 1 label for the whole batch of images. I tried with duplicating targets to be the same size as input but that also didn't work.

**Comment by diksham-flwls** [comment]
To test things out I ran a small example on resnet with a batch of images but I still get a single cam output for a batch of images?
```
from pytorch_grad_cam import GradCAM, HiResCAM, ScoreCAM, GradCAMPlusPlus, AblationCAM, XGradCAM, EigenCAM, FullGrad
from pytorch_grad_cam.utils.model_targets import ClassifierOutputTarget
from pytorch_grad_cam.utils.image import show_cam_on_image
from torchvision.models import resnet50

import torchvision.transforms as transforms
from PIL import Image

invTrans = transforms.Compose([ transforms.Normalize(mean = [ 0., 0., 0. ],
                                                     std = [ 1/0.26862954, 1/0.26130258, 1/0.27577711 ]),
                                transforms.Normalize(mean = [-0.48145466, -0.4578275, -0.40821073],
                                                     std = [ 1., 1., 1. ]),
                               ])

def process_img(img_path):
    try:
        img = Image.open(img_path).convert(mode="RGB")
    except Exception as err:
        print(f"Unexpected {err=}, {type(err)=} while trying to read file:{img_path}")

    transform = transforms.Compose([
        transforms.RandomChoice([
            transforms.ColorJitter(saturation=0.5),
            transforms.ColorJitter(contrast=0.5),
            transforms.GaussianBlur(kernel_size=(5, 9), sigma=(0.1, 5)),
            # AddGaussianNoise(mean=0.0, std=0.1),
        ]),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.48145466, 0.4578275, 0.40821073],
                                    std=[0.26862954, 0.26130258, 0.27577711]),
        transforms.Resize((224, 224))
    ])
    img = transform(img)
    return img

model = resnet50(pretrained=True)
target_layers = [model.layer4[-1]]
real_sample = "./fish.JPEG"
img = process_img(real_sample)
input_tensor = img.unsqueeze(dim=0).tile(dims=(3,1,1,1)) # Create an input tensor image for your model..

# Construct the CAM object once, and then re-use it on many images:
cam = GradCAM(model=model, target_layers=target_layers)

targets = [ClassifierOutputTarget(0)] * 3

grayscale_cam = cam(input_tensor=input_tensor, targets=targets)

print(grayscale_cam.shape)

```

**Comment by Lotto98** [comment]
i'm facing the same problem. I have an input tensor of (BATCH, CHANNELS, 224, 224) but GradCAM returns only a (224, 224) array. I was expecting a (BATCH, 224, 224) array. Did you manage to understand why this is happening?

**Comment by diksham-flwls** [comment]
Well in my case I had to transform my input into a batch because my input was a tuple of images. However the target was a single element. GradCAM computes the class activation map with respect to the given target. 
If you give a batch target it should give you a batch output of GradCAM.

**Comment by Lotto98** [comment]
My `target` is `[ClassifierOutputTarget(0), ClassifierOutputTarget(1)]` and my `input tensor` is a batch of 2 images and has size [2, 3, 224, 224]. Thus I'm expecting a [2, 240, 240] array as GradCAM output, but the returned size is [240, 240].

**Comment by Lotto98** [comment]
My bad! I found the error. I was reading always the first element of the list in the function which calls GradCam on my batch. Sorry for wasting  your time.

**Comment by jacobgil** [comment]
Hi @diksham-flwls sorry for the late reply,
If I understand correctly, this isn't really a batch - the model operates on all these at once and outputs a single prediction, and not one prediction per batch member, is that correct ?

If it is, then a workaround that could be done, would be to create a wrapper that fixes all the input images except the i'th image, and then compute grad-cam for it.

**Comment by diksham-flwls** [comment]
@jacobgil Thank you for getting back to me. 
Yes I ended up creating a wrapper that would take that batch of images and reshape it in the format expected by the model. Also my model expected a tuple of tensors with different shapes that could not be merged into 1 batch tensor so I ended up moving it out of the forward call as shown below:

```
class ToyModelV2(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
    
    def set_features(self, img):
        features = self.model.get_features(img)
        self.features = features
    
    def forward(self, input):                                                              # n_crops * n_time x 3 x 224 x 224
        input = input.reshape(3, 5, -1, input.shape[1], input.shape[2], input.shape[3])    # n_crops x n_time x B x 3 x 224 x 224
        crops = [ [frame for frame in crop] for crop in input ]
        out1, _, _, = self.model(crops, self.features)
        return out1
```
